File: project2/src/game_state/states.py
from dataclasses import dataclass
from typing import List
from src.poker_oracle.deck import Card
from src.game_manager.game_stage import GameStage


@dataclass
class PlayerState:
    def __init__(self, chips: int, folded: bool, bet: int):
        self.chips = chips  # An integer
        self.folded = folded  # A boolean
        self.bet = bet  # An integer


@dataclass
class BoardState:
    def __init__(self, cards: List[Card], pot: int, highest_bet: int, dealer: int):
        if not isinstance(cards, list):
            raise TypeError("cards must be a list")
        self.cards = cards  # A list of Card instances
        self.pot = pot  # An integer
        self.highest_bet = highest_bet  # An integer
        self.dealer = dealer  # An integer


@dataclass
class GameState:
    def __init__(
        self,
        player_states: List[PlayerState],
        board_state: BoardState,
        game_stage: GameStage,
    ):
        self.player_states = player_states  # A list of PlayerState instances
        self.board_state = board_state  # An instance of BoardState
        self.game_stage = game_stage  # An instance of GameStage

File: project2_rust/src/game_state/states.rs
use crate::card::Card;
use crate::game_manager::GameStage;
pub struct PlayerState {
    pub chips: u32,
    pub folded: bool,
    pub bet: u32,
}
impl PlayerState {
    pub fn new(chips: u32, folded: bool, bet: u32) -> Self {
        Self { chips, folded, bet }
    }
}

pub struct BoardState {
    pub cards: Vec<Card>,
    pub pot: u32,
    pub highest_bet: u32,
    pub dealer: usize,
}

impl BoardState {
    pub fn new(cards: Vec<Card>, pot: u32, highest_bet: u32, dealer: usize) -> Self {
        Self {
            cards,
            pot,
            highest_bet,
            dealer,
        }
    }
}

pub struct GameState {
    player_states: Vec<PlayerState>,
    board_state: BoardState,
    game_stage: GameStage,
}

impl GameState {
    // TODO: Implement player bid history
    pub fn new(
        player_states: Vec<PlayerState>,
        board_state: BoardState,
        game_stage: GameStage,
    ) -> Self {
        Self {
            player_states,
            board_state,
            game_stage,
        }
    }
}
